File: src/v8/experts/floor_trader_pivot.py
"""Floor-trader daily-pivot S/R reaction behavior family (`floor_trader_pivot`).

Hypothesis (mechanism `pivot_level_reaction`; Ch8.7 p263-265, Ch10.13): the
daily pivot level set (PP/S1..S4/R1..R4 from the prior period's OHLC) is a
time-static support/resistance ladder, and a close-confirmed reaction at a
level is a directional entry. Detected on closed bars only; entry is
NEXT_BAR_CLOSE; confirmation is close-based (Ch14.2).

Variants (D-044; the card's variant e — pivot levels as TARGETS — is folded
into the target ladder of a-d, not a separate entry):
  a  PP drift rule (Ch8.7 p264): open above PP and close up -> LONG drift
     (target R1); open below PP and close down -> SHORT (target S1).
  b  S1/R1 range-day reaction: a pullback to S1 that closes back above it
     (LONG, target PP) / a rally to R1 that closes back below it (SHORT).
  c  S2/R2 violation = strong trend (Ch8.7 p264): a close beyond the second
     support/resistance is continuation -> SHORT below S2 (target S3) /
     LONG above R2 (target R3).
  d  S3/R3 extreme reversion: a close beyond the third level that reclaims
     it is an extreme overbought/oversold fade -> LONG reclaim above S3
     (target S2) / SHORT reclaim below R3 (target R2).

Stop: beyond the level being traded (behind the pivot line, Ch8.7 doctrine).
Target: the next level on the ladder (the level set IS the target ladder,
Ch8.7 p263-265). Targets/stops are price distances expressed in R with the
14-bar ATR unit (D-028).

The level set is FROZEN at detection and the anchor is restricted to the
current session: the pivot set is recomputed daily and a reaction in a later
day is a different setup (D-026: a new anchor event). `still_valid` keeps the
thesis alive while the close stays on the traded side of the frozen level.
"""
from __future__ import annotations

from ..schema import MarketState, CandidateDraft, ExpertEvaluation
from .base import Expert


class FloorTraderPivotExpert(Expert):
    """Daily-pivot level reaction expert."""
    expert_id = 'floor_trader_pivot'
    version = 'v1'
    mechanism_family_id = 'pivot_level_reaction'
    behavior_family_id = 'pivot_level_reaction'
    variant_id = 'a'
    # D-044: every implemented variant (losers included); the reported
    # variant_id is a member. D-046: all thresholds are declared constants
    # frozen pre-window, so the search universe equals the evaluated set.
    variants_evaluated = ('a', 'b', 'c', 'd')
    search_universe_size = 4
    requires = ('location', 'volatility', 'history', 'session')

    def __init__(self, variant_id: str | None = None):
        if variant_id is not None:
            if variant_id not in self.variants_evaluated:
                raise ValueError(
                    f'{self.expert_id}: unknown variant {variant_id!r} '
                    f'(variants_evaluated={list(self.variants_evaluated)})')
            self.variant_id = variant_id

    def _detect(self, f: dict, n: int) -> tuple | None:
        """(direction, level_ref, stop_price, target_price, anchor_pred) for
        the configured variant on the current bar; None = no reaction."""
        ppv = f[f'{self._sym}.pivot_points_day'].value
        PP, R1, R2, R3, _R4, S1, S2, S3, _S4 = (float(v) for v in ppv)
        o = float(self._hist[n - 1][1])
        h = float(self._hist[n - 1][2])
        l = float(self._hist[n - 1][3])
        c = float(self._hist[n - 1][4])
        # The anchor is restricted to the current session: the level set is
        # recomputed every day, so a reaction in a later day is a new setup
        # (D-026 anchor = the market event that created the setup).
        day_start = n - int(self._bar_of_session)
        v = self.variant_id
        if v == 'a':
            if o > PP and c > o:
                pred = lambda j, bar: j >= day_start \
                    and float(bar[1]) > PP and float(bar[4]) > float(bar[1])
                return 'LONG', PP, PP, R1, pred
            if o < PP and c < o:
                pred = lambda j, bar: j >= day_start \
                    and float(bar[1]) < PP and float(bar[4]) < float(bar[1])
                return 'SHORT', PP, PP, S1, pred
        elif v == 'b':
            if l <= S1 and c > S1:
                pred = lambda j, bar: j >= day_start \
                    and float(bar[3]) <= S1 and float(bar[4]) > S1
                return 'LONG', S1, S1, PP, pred
            if h >= R1 and c < R1:
                pred = lambda j, bar: j >= day_start \
                    and float(bar[2]) >= R1 and float(bar[4]) < R1
                return 'SHORT', R1, R1, PP, pred
        elif v == 'c':
            if c > R2:
                pred = lambda j, bar: j >= day_start and float(bar[4]) > R2
                return 'LONG', R2, R2, R3, pred
            if c < S2:
                pred = lambda j, bar: j >= day_start and float(bar[4]) < S2
                return 'SHORT', S2, S2, S3, pred
        else:  # 'd' — S3/R3 extreme reversion
            if l <= S3 and c > S3:
                pred = lambda j, bar: j >= day_start \
                    and float(bar[3]) <= S3 and float(bar[4]) > S3
                return 'LONG', S3, S3, S2, pred
            if h >= R3 and c < R3:
                pred = lambda j, bar: j >= day_start \
                    and float(bar[2]) >= R3 and float(bar[4]) < R3
                return 'SHORT', R3, R3, R2, pred
        return None

    def evaluate(self, state: MarketState) -> ExpertEvaluation:
        t = state.as_of
        sym = state.universe[0]
        need = [f'{sym}.close', f'{sym}.atr', f'{sym}.history',
                f'{sym}.pivot_points_day', f'{sym}.bar_of_session']
        if not self._need(state, need):
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        f = state.features
        close = float(f[f'{sym}.close'].value)
        atr = f[f'{sym}.atr'].value
        hist_value = f[f'{sym}.history'].value
        if not isinstance(hist_value, (tuple, list)) or not hist_value or atr is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        self._hist = tuple(hist_value)
        self._sym = sym
        self._bar_of_session = float(f[f'{sym}.bar_of_session'].value)
        n = len(self._hist)
        if n < 2:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        hit = self._detect(f, n)
        if hit is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        direction, level, stop_price, target_price, pred = hit
        if direction == 'LONG':
            stop_r = (close - stop_price) / atr
            target_r = (target_price - close) / atr
        else:
            stop_r = (stop_price - close) / atr
            target_r = (close - target_price) / atr
        if stop_r <= 0 or target_r <= 0:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        anchor = self.find_setup_anchor(self._hist, pred)
        geometry = {'entry': 'NEXT_BAR_CLOSE', 'target_r': target_r,
                    'stop_r': stop_r, 'expiry_bars': 8, 'atr_ref': atr,
                    'variant': self.variant_id,
                    'level_ref': level, 'stop_ref': stop_price}
        if direction == 'LONG':
            geometry['prior_low_ref'] = stop_price
        else:
            geometry['prior_high_ref'] = stop_price
        draft = CandidateDraft(
            expert_id=self.expert_id, expert_version=self.version,
            instrument=sym, direction=direction,
            setup_fingerprint=(f'{sym}:{self.variant_id}:{direction}:'
                               f'{close:.6f}:{level:.6f}:{stop_price:.6f}'),
            risk_geometry=geometry,
            birth_time=t, setup_anchor_event_id=anchor)
        return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                'APPLICABLE', 'CANDIDATE', t, draft)

    def still_valid(self, state: MarketState, draft: CandidateDraft) -> bool:
        """The thesis is "the reaction holds on the traded side of the level":
        a close back through the FROZEN level says the reaction failed (role
        reversal applies, Ch8.7)."""
        sym = draft.instrument
        f = state.features
        close = f.get(f'{sym}.close')
        if close is None or close.value is None:
            return True          # unobservable: fail open, price still governs
        ref = draft.risk_geometry.get('level_ref')
        if ref is None:
            return True
        if draft.direction == 'LONG':
            return float(close.value) > float(ref)
        return float(close.value) < float(ref)

File: v8-core/src/experts/floor_trader_pivot.rs
//! floor_trader_pivot: daily-pivot S/R reaction, default variant 'a' (the
//! PP-drift rule, Ch8.7 p264). Port of src/v8/experts/floor_trader_pivot.py
//! evaluate() bit-for-bit (PARITY_AND_IDENTITY_SPEC §3). The D-026 anchor is
//! restricted to the current session (day_start = n - bar_of_session): the
//! pivot set is recomputed daily, so a reaction in a later day is a new
//! setup. The parity harness constructs the Python expert with its default
//! variant ('a'), so only that variant is ported here.

use crate::experts::base::*;
use crate::simulator::Draft;
use crate::state::HistBar;

pub const PORTED: bool = true;
pub const VERSION: &str = "v1";
pub const REQUIRES: &[&str] = &["location", "volatility", "history", "session"];
// Declared risk geometry (EXPERT_PROTOCOL §1: risk geometry is "Predeclared
// entry, stop, target, timeout and sizing inputs"; SIMULATION_TRUTH_SPEC D-028:
// R is a declared price distance). Fixed values are declared here, never
// re-literalized inside evaluate(); a structural target/stop is computed at
// the call site and overrides the key.
pub const EXPIRY_BARS: i64 = 8;

pub fn floor_trader_pivot(fm: &FeatMap, expert_id: &str, version: &str) -> ExpertEval {
    let sym = fm.symbol;
    // Python `_need`: {sym}.close, {sym}.atr, {sym}.history,
    // {sym}.pivot_points_day, {sym}.bar_of_session — the FeatMap is keyed by
    // the bare names.
    let close = match fm.value("close") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let atr = match fm.value("atr") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    if fm.history.is_empty() {
        return no_habitat(expert_id, version, fm.as_of);
    }
    // pivot_points_day is a non-scalar feature: (PP, R1, R2, R3, R4, S1, S2,
    // S3, S4), the Python unpack order.
    let ppv = match fm.features.get("pivot_points_day") {
        Some(f) => match f.value.as_array() {
            Some(a) => a,
            None => return no_habitat(expert_id, version, fm.as_of),
        },
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    if ppv.len() < 9 {
        return no_habitat(expert_id, version, fm.as_of);
    }
    let fv = |i: usize| ppv[i].as_f64().unwrap();
    let pp = fv(0);
    let r1 = fv(1);
    let _r2 = fv(2);
    let _r3 = fv(3);
    let _r4 = fv(4);
    let s1 = fv(5);
    let _s2 = fv(6);
    let _s3 = fv(7);
    let _s4 = fv(8);
    let bos = match fm.value("bar_of_session") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let n = fm.history.len();
    if n < 2 {
        return no_habitat(expert_id, version, fm.as_of);
    }
    // _detect on the newest history bar (bar t-1).
    let o = fm.history[n - 1].open;
    let _h = fm.history[n - 1].high;
    let _l = fm.history[n - 1].low;
    let c = fm.history[n - 1].close;
    // int(bar_of_session) truncates toward zero; the feature is an
    // integer-valued float, so the `as i64` cast is exact.
    let day_start = n as i64 - bos as i64;
    // Variant 'a' only (the Python default the harness evaluates).
    let pred: Box<dyn Fn(usize, &HistBar) -> bool>;
    let (direction, level, stop_price, target_price): (String, f64, f64, f64);
    if o > pp && c > o {
        direction = "LONG".into();
        level = pp;
        stop_price = pp;
        target_price = r1;
        pred = Box::new(move |j, b| (j as i64) >= day_start && b.open > pp && b.close > b.open);
    } else if o < pp && c < o {
        direction = "SHORT".into();
        level = pp;
        stop_price = pp;
        target_price = s1;
        pred = Box::new(move |j, b| (j as i64) >= day_start && b.open < pp && b.close < b.open);
    } else {
        return no_setup(expert_id, version, fm.as_of);
    }
    let (stop_r, target_r) = if direction == "LONG" {
        ((close - stop_price) / atr, (target_price - close) / atr)
    } else {
        ((stop_price - close) / atr, (close - target_price) / atr)
    };
    if stop_r <= 0.0 || target_r <= 0.0 {
        return no_setup(expert_id, version, fm.as_of);
    }
    let anchor = find_setup_anchor(&fm.history, &*pred);
    let mut geometry = geom(vec![
        ("entry", serde_json::json!("NEXT_BAR_CLOSE")),
        ("target_r", serde_json::json!(target_r)),
        ("stop_r", serde_json::json!(stop_r)),
        ("expiry_bars", serde_json::json!(EXPIRY_BARS)),
        ("atr_ref", serde_json::json!(atr)),
        ("variant", serde_json::json!("a")),
        ("level_ref", serde_json::json!(level)),
        ("stop_ref", serde_json::json!(stop_price)),
    ]);
    if direction == "LONG" {
        geometry.insert("prior_low_ref".to_string(), serde_json::json!(stop_price));
    } else {
        geometry.insert("prior_high_ref".to_string(), serde_json::json!(stop_price));
    }
    let fingerprint = format!(
        "{sym}:a:{direction}:{:.6}:{:.6}:{:.6}",
        close, level, stop_price
    );
    let draft = Draft {
        direction: direction.clone(),
        birth_time: fm.as_of,
        risk_geometry: geometry,
    };
    candidate(expert_id, version, fm.as_of, draft, anchor, fingerprint)
}
